### ISSUE TITLE
Help transpiling to ES5 for MongoDB Realm

### ISSUE THREAD

**Issue by caschbre** [sveltech/ssr#3]
I'm attempting to get routify running on [MongoDB realm](https://docs.mongodb.com/realm/) using SSR. After reviewing how to get SSR working with Netlify and Now, I need to get the api/ssr.js script to run in a [MongoDB Realm function](https://docs.mongodb.com/realm/functions/). Since the ssr.js script has dependencies on @sveltech/ssr and fs-extra, MongoDB Realm needs those packages [uploaded first](https://docs.mongodb.com/realm/functions/upload-external-dependencies/) so they can be imported in the function script.

According to the upload dependencies documentation, "MongoDB Realm automatically transpiles uploaded modules to ES5 and supports most built-in modules."  This seems to fail when @sveltech/ssr is the dependent node module.

Any suggestions / pointers?  I feel MongoDB Realm would be a great fit for Routify (which runs great as an SPA), however I'm stuck on getting SSR working.

**Comment by rixo** [comment]
> This seems to fail when @sveltech/ssr is the dependent node module.

Do you have details on how it fails? Stack trace?

**Comment by caschbre** [comment]
Unfortunately I haven't been able to find those details on MongoDB Realm yet. When uploading the compressed node modules folder, per the docs, I get a cryptic error:

`Failed to upload node_modules.tar.gz: unknown: Unexpected token, expected ( (357:12) 355 | 356 | return true; > 357 | } catch { | ^ 358 | return false; 359 | } 360 | }`

And when I try to import it through their realm-cli it gives me even less info... and there isn't a verbose option. ugh.

I tried using rollup to bundle the ssr.js and dependencies all into one file and transpile down to ES5 but I was receiving errors locally when doing that, however I may have been shooting in the dark attempting to do that.

**Comment by jakobrosenberg** [comment]
Unexpected token in a compressed file during upload? I have no idea what to make of that.

Could you check if there's a dependency for `canva` somewhere?

**Comment by caschbre** [comment]
I believe the unexpected token happens when the compressed node modules folder is uploaded and MongoDB Realm attempts to transpile it.

> Could you check if there's a dependency for canva somewhere?

Do you mean canvas?  There is a dependency for canvas in jsdom I believe.

I was trying to debug locally and also came across an unexpected token error. I initially though I was just doing it wrong but it may actually be the same issue. Here's a gist of the test I was trying locally...

https://gist.github.com/caschbre/80dfcc191f7539edaea2e7495769e3a9

**Comment by jakobrosenberg** [comment]
Have you tried adding `@rollup/plugin-json`? I'm not sure why it would be needed, but the gist suggests you do.

**Comment by caschbre** [comment]
Yep, that was one of the first things I tried but it didn't change the outcome.

**Comment by caschbre** [comment]
So further local debugging leads me to believe this might be a jsdom issue?

I tracked the unexpected token issue to a [statement in jsdom](https://github.com/jsdom/jsdom/blob/master/lib/jsdom/browser/resources/resource-loader.js#L7) that is doing a `const packageVersion = require("../../../../package.json").version;`  If I edit that line and make it `const packageVersion = "1.2.3";` then I no longer see that specific unexpected token issue.  However, it moves on to the next unexpected token issue with is another `const foobar = required('./somefile.json')`

Does that make any sense?  Any suggestions on working around that?

**Comment by caschbre** [comment]
Well, I take that back.... it does seem to be the require call on json files, but it's not limited to jsdom.

**Comment by jakobrosenberg** [comment]
You could possibly use Rollup to compile a dist for MongoDB.

```javascript
   const rollupConfig = {
        input: `mongo-realm/ssr.js`,
        output: {
            name: 'mongodb_realm',
            sourcemap: true,
            format: 'iife',
            file: `mongo-realm/api/ssr.js`
        },
        plugins: [
            commonjs(),
            // resolve({ browser: true }), //may need to uncomment this line
            production && terser(),
            replace({ 'process.env.NODE_ENV': "'production'" })
        ]
    }
```

**Comment by caschbre** [comment]
Thanks! I'll give that a shot. Initial test does build.

**Comment by caschbre** [comment]
So after more fiddling I did end up finding a way to use @rollup/plugin-json... with the idea that I don't need to upload the compressed node_modules folder since it fails transpiling.  It does generate a 3MB minimized js file. ugh.

https://gist.github.com/caschbre/743d671a4623d1e23f07ae74885bdeb5

The idea is to then take that generated file and use that as the MongoDB function.  I haven't figured out yet how to handle routing to that function similar to now/netlify.  I don't think MongoDB offers that.  My guess is that I can setup the [MongoDB hosting for SPA](https://docs.mongodb.com/realm/hosting/host-a-single-page-application/).

I probably need to make an index.js file using one of the Realm SDKs and then have that make a call to the serverless function. Ugh.

So at this point we can probably close this ticket until I have time to further investigate making MongoDB Realm work. I'm a bit burnt out now. lol.

**Comment by jakobrosenberg** [comment]
You should be able to grab the request somehow and pass the path to `ssr`.

**Comment by caschbre** [comment]
I believe so. When setting up [MongoDB hosting for SPA](https://docs.mongodb.com/realm/hosting/host-a-single-page-application/) you can configure it to route all requests to a single (hosted) file.  I'm guessing a simple index.js file that could read the incoming request / path.  The next step would then be using one of the Realm SDKs to call the ssr serverless function.

Another option that might be simpler is to setup a webhook on Realm. Then index.js could read the path and make another call to the webhook passing in the path. The webhook function already has access to make the serverless function call.

And both of those assume that the 3mb ssr.js file works as a function.  I haven't yet determined whether the serverless function has access to the hosted files containing the __app.html file.

Ugh... while typing this I tried uploading the generated ssr file but Realm limits serverless function source to 100000 characters.

**Comment by caschbre** [comment]
Received more info from MongoDB regarding the transpiling issue....

> We took a look and the reason for this error is that the package depends on a version of babel that is higher than the version of babel we use internally to transpile the source code. We are using babel 6.26.3. Can you use a lower version ?

I'm not sure that helps any but thought you might have more insight.

**Comment by jakobrosenberg** [comment]
Do you know which package depends on babel? Routify itself shouldn't have any babel requirements.

**Comment by caschbre** [comment]
I did an `npm -i` with just @sveltech/ssr in the package.json and then scanned the node_modules folder... the only packages that had babel in their package.json files were....

```
node_modules/node-fetch/package.json
node_modules/estraverse/package.json
node_modules/cssstyle/package.json
node_modules/symbol-tree/package.json
node_modules/uri-js/package.json
```

I then added `"babel-core": "6.26.3"` to the package.json and rebuild the node_modules folder. That still failed when attempting to upload it to MongoDB Realm so i'm guessing that's not the issue, but i'm still working through.

I don't think this is an issue specifically with sveltech/ssr but I'm just keeping this issue updated with what I find in case something jumps out to you or if anyone else is trying to attempt this.

FYI... here's the thread I have open on the MongoDB community forums.
https://developer.mongodb.com/community/forums/t/failed-to-upload-node-modules-tar-gz-unknown-unexpected-token/6228
